File: packages/core/src/i18n.ts
// Shared, framework-agnostic i18n core. The JSON catalogs in ./locales are the
// single source of truth for every KROMA surface they are bundled into the TS
// clients here AND `include_str!`'d by the Rust server (see server/src/i18n.rs),
// so message keys stay in lockstep across the whole stack.
//
// React bindings (provider + hooks) live in `@kroma/ui` to keep this package
// React-free; the server has its own tiny mirror of `translate`.

import en from './locales/en.json';
import fr from './locales/fr.json';

/** Supported UI locales. `fr` is the project default. */
export type Locale = 'fr' | 'en';

/** Every translatable key. `fr.json` is the authoritative key set; other locales
 * may lag and fall back to French (then to the raw key). */
export type MessageKey = keyof typeof fr;

/** Interpolation values for `{placeholder}` tokens in a message. */
export type TVars = Record<string, string | number>;

/** A bound translation function for one locale. */
export type Translate = (key: MessageKey, vars?: TVars) => string;

export const DEFAULT_LOCALE: Locale = 'fr';

/** Ordered list of selectable locales, with the key for their native label. */
export const LOCALES: ReadonlyArray<{ code: Locale; labelKey: MessageKey }> = [
  { code: 'fr', labelKey: 'lang.fr' },
  { code: 'en', labelKey: 'lang.en' },
];

const CATALOGS: Record<Locale, Record<string, string>> = { fr, en };

/** Narrow an arbitrary value to a supported {@link Locale}. */
export function isLocale(value: unknown): value is Locale {
  return value === 'fr' || value === 'en';
}

/** Map a BCP-47 tag (`"en-US"`, `"fr"`, `"FR"`) to a supported locale, or null
 * when none matches. Also accepts the server's display names (`"Français"`). */
export function normalizeLocale(tag?: string | null): Locale | null {
  if (!tag) return null;
  const base = tag.toLowerCase().split(/[-_]/)[0];
  if (base === 'fr' || tag === 'Français') return 'fr';
  if (base === 'en' || tag === 'English') return 'en';
  return null;
}

/** Best locale for the current device: an explicit preference wins, then the
 * browser's languages, else the default. Safe to call without a DOM. */
export function detectLocale(preferred?: string | null): Locale {
  const explicit = normalizeLocale(preferred);
  if (explicit) return explicit;
  const nav = typeof navigator !== 'undefined' ? navigator : undefined;
  let tags: readonly string[] = [];
  if (nav?.languages?.length) tags = nav.languages;
  else if (nav?.language) tags = [nav.language];
  for (const tag of tags) {
    const loc = normalizeLocale(tag);
    if (loc) return loc;
  }
  return DEFAULT_LOCALE;
}

/** Replace `{name}` tokens in `template` from `vars`. Unknown tokens are kept. */
function interpolate(template: string, vars?: TVars): string {
  if (!vars) return template;
  return template.replace(/\{(\w+)\}/g, (whole, name: string) =>
    name in vars ? String(vars[name]) : whole,
  );
}

/** A single-locale message catalog (dotted key -> template). */
export type Catalog = Record<string, string>;
/** A per-locale set of catalogs (some locales may be absent). */
export type Catalogs = Partial<Record<Locale, Catalog>>;

/** Whether a key resolves in `catalogs` for `locale` (or the French fallback). */
function hasKeyIn(catalogs: Catalogs, locale: Locale, key: string): boolean {
  return catalogs[locale]?.[key] != null || catalogs[DEFAULT_LOCALE]?.[key] != null;
}

/**
 * i18next-style plural resolution over an explicit catalog set. When a
 * translation is called with a numeric `count`, the catalog can carry CLDR plural
 * variants suffixed with the category name `key_one`, `key_other` (and
 * `_zero`/`_two`/`_few`/`_many` where a locale needs them). `Intl.PluralRules`
 * picks the category; we use the matching variant, else `key_other`, else the
 * base key. `{count}` is available as an interpolation token in every variant.
 *
 *   "content.seasonCount":      "{count} saisons"   ← base / default
 *   "content.seasonCount_one":  "{count} saison"    ← used when count selects "one"
 *   t('content.seasonCount', { count })             ← call site stays count-only
 */
function pluralKeyIn(catalogs: Catalogs, locale: Locale, key: string, count: number): string {
  let category: Intl.LDMLPluralRule = count === 1 ? 'one' : 'other';
  try {
    category = new Intl.PluralRules(locale).select(count);
  } catch {
    /* environments without Intl.PluralRules → the one/other heuristic above */
  }
  const variant = `${key}_${category}`;
  if (hasKeyIn(catalogs, locale, variant)) return variant;
  const other = `${key}_other`;
  if (hasKeyIn(catalogs, locale, other)) return other;
  return key;
}

/** Translate `key` against an explicit catalog set (e.g. a module's own
 * catalogs), returning `undefined` when the key is absent so the caller can fall
 * back to the core translator. Applies the same plural + `{name}` interpolation
 * as {@link translate}. */
export function translateIn(
  catalogs: Catalogs,
  locale: Locale,
  key: string,
  vars?: TVars,
): string | undefined {
  const lookupKey =
    typeof vars?.count === 'number' ? pluralKeyIn(catalogs, locale, key, vars.count) : key;
  const template = catalogs[locale]?.[lookupKey] ?? catalogs[DEFAULT_LOCALE]?.[lookupKey];
  return template == null ? undefined : interpolate(template, vars);
}

/** Translate a key in a locale, falling back to French then to the raw key.
 * Pass a numeric `count` in `vars` to select a plural variant; `{count}` and any
 * other `vars` are interpolated. */
export function translate(locale: Locale, key: MessageKey, vars?: TVars): string {
  return translateIn(CATALOGS, locale, key, vars) ?? key;
}

/** Build a translation function bound to `locale`. */
export function createTranslator(locale: Locale): Translate {
  return (key, vars) => translate(locale, key, vars);
}

File: server/src/i18n.rs
//! Server-side i18n a tiny mirror of the shared `@luma/core` i18n.
//!
//! The JSON catalogs in `packages/core/src/locales` are the single source of
//! truth for the whole stack (they are also bundled into the TS clients). We
//! `include_str!` them at compile time and apply the same `{var}` interpolation,
//! so a key added on the client is automatically available on the server.

use std::collections::HashMap;
use std::convert::Infallible;
use std::sync::OnceLock;

use axum::async_trait;
use axum::extract::FromRequestParts;
use axum::http::request::Parts;

use crate::state::SharedState;

/// Default locale (matches the TS `DEFAULT_LOCALE`).
pub const DEFAULT_LOCALE: &str = "fr";

const FR_JSON: &str = include_str!("../../packages/core/src/locales/fr.json");
const EN_JSON: &str = include_str!("../../packages/core/src/locales/en.json");

/// Lazily-parsed catalog for a locale (falls back to French for unknown codes).
fn catalog(locale: &str) -> &'static HashMap<String, String> {
    static FR: OnceLock<HashMap<String, String>> = OnceLock::new();
    static EN: OnceLock<HashMap<String, String>> = OnceLock::new();
    match locale {
        "en" => EN.get_or_init(|| serde_json::from_str(EN_JSON).expect("en.json catalog")),
        _ => FR.get_or_init(|| serde_json::from_str(FR_JSON).expect("fr.json catalog")),
    }
}

/// Map a BCP-47 tag, an `Accept-Language` header, or one of the server's display
/// names (`"Français"`/`"English"`) to a supported locale code, or `None`.
pub fn normalize(tag: &str) -> Option<&'static str> {
    match tag.trim() {
        "Français" => return Some("fr"),
        "English" => return Some("en"),
        _ => {}
    }
    // Take the highest-priority listed language and strip any region/quality
    // suffix: "en-US,en;q=0.9,fr;q=0.8" → "en", "fr-CH" → "fr".
    let base = tag
        .split([',', ';', '-', '_'])
        .next()
        .unwrap_or("")
        .trim()
        .to_ascii_lowercase();
    match base.as_str() {
        "fr" => Some("fr"),
        "en" => Some("en"),
        _ => None,
    }
}

/// Replace `{name}` tokens in `template` from `vars`. Unknown tokens are kept.
fn interpolate(template: &str, vars: &[(&str, &str)]) -> String {
    if vars.is_empty() {
        return template.to_string();
    }
    let mut out = template.to_string();
    for (name, value) in vars {
        out = out.replace(&format!("{{{name}}}"), value);
    }
    out
}

/// Translate `key` in `locale`, falling back to French then to the raw key.
pub fn t(locale: &str, key: &str, vars: &[(&str, &str)]) -> String {
    let template = catalog(locale)
        .get(key)
        .or_else(|| catalog(DEFAULT_LOCALE).get(key))
        .map(String::as_str)
        .unwrap_or(key);
    interpolate(template, vars)
}

/// The resolved request locale. Clients send their active locale (the account
/// preference once signed in) as `Accept-Language`; we honour it, else fall back
/// to the default. Drives every server-rendered string (admin settings labels,
/// error messages).
pub struct ReqLocale(pub &'static str);

#[async_trait]
impl FromRequestParts<SharedState> for ReqLocale {
    type Rejection = Infallible;

    async fn from_request_parts(
        parts: &mut Parts,
        _state: &SharedState,
    ) -> Result<Self, Self::Rejection> {
        let locale = parts
            .headers
            .get(axum::http::header::ACCEPT_LANGUAGE)
            .and_then(|v| v.to_str().ok())
            .and_then(normalize)
            .unwrap_or(DEFAULT_LOCALE);
        Ok(ReqLocale(locale))
    }
}
